Editorial suggestions from Benjamin Kaduk (invariants, TLS) (#4473)
quicwg/base-drafts: draft-ietf-quic-tls.md

~~~~
          +-------------+------------+--------------+---------+

> Content   |             |            |  Application |         |

Reviewer: @ekr, do you have any opinion on this?

the process by providing TLS with the client's handshake bytes.

At any time, the TLS stack at an endpoint will have a current sending

> encryption level and receiving encryption level. TLS Encryption levels determine

Reviewer: ```suggestion
encryption level and receiving encryption level. TLS encryption levels determine
```

about the current connection when creating a resumed connection. TLS requires
that some information be retained; see Section 4.6.1 of {{!TLS13}}. QUIC itself
does not depend on any state being retained when resuming a connection, unless

> 0-RTT is also used; see {{enable-0rtt}} and Section 7.4.1 of

Reviewer 2: This might even warrant an "...of this document" to make it explicit.

### Enabling 0-RTT {#enable-0rtt}

The TLS "early_data" extension in the NewSessionTicket message is defined
to convey (in the "max_early_data_size" parameter) the amount of TLS 0-RTT

> data the server is willing to accept.  QUIC does not use TLS 0-RTT data;
> instead, QUIC has its own 0-RTT stream.  Accordingly, the

Reviewer: ```suggestion
data the server is willing to accept.  QUIC does not use TLS 0-RTT data.
QUIC uses 0-RTT packets to carry early data.  Accordingly, the
```

"max_early_data_size" parameter is repurposed to hold a sentinel value
0xffffffff to indicate that the server is willing to accept QUIC 0-RTT data;
to indicate that the server does not accept 0-RTT data, the "early_data"

> extension is omitted from the NewSessionTicket.

Author: (This all might become moot depending on the resolution of my discuss point.)

## Keys

> The labels (Hkdf.label) generated during the execution of the HKDF-Expand-Label function and

Reviewer: I don't think that this is right.  I think that you mean:

```suggestion
The labels generated during the execution of the HKDF-Expand-Label function (that is, HkdfLabel.label) and part of the value
```

Author: Oops, yes, the HkdfLabel bit got garbled in transmission from brain to fingers.
Your suggestion looks fine; I just wanted to reiterate that it's the whole HkdfLabel.label (including the "tls13 " prefix) that we show, not just the argument to HKDF-Expand-Label.

the protected payload is sampled, then applied to the header:

~~~

> sampled_data = d1b1c98dd7689fb8ec11d242b123dc9b

Reviewer: ```suggestion
sampled_bytes = d1b1c98dd7689fb8ec11d242b123dc9b
```

> mask = AES-ECB(hp, sampled_data)[0..4]

Reviewer: ```suggestion
mask = AES-ECB(hp, sampled_bytes)[0..4]
```

from the third protected byte:

~~~

> sampled_data = 2cd0991cd25b0aac406a5816b6394100

Reviewer: ```suggestion
sampled_bytes = 2cd0991cd25b0aac406a5816b6394100
```

produce the sample for header protection.

~~~

> sampled_data = 5e5cd55c41f69080575d7999c25a5bfb

Reviewer: ```suggestion
sampled_bytes = 5e5cd55c41f69080575d7999c25a5bfb
```
